### PR TITLE
[Core] Add reverse and random scenario execution order

### PR DESCRIPTION
**Pull request by grasshopper7**
## Summary

Allows scenarios to be executed in random or reverse order. Custom execution orders can also be created.

## Details

All scenarios can be run in random order. 
`java <classpath> cucumber.api.cli.Main --order random -g <glue path> <feature file path>`
The number of scenarios to be run in random can also be mentioned.
`java <classpath> cucumber.api.cli.Main --order random --count <number> -g <glue> <feature file>`

All scenarios can be run in reverse order.
`java <classpath> cucumber.api.cli.Main --order reverse -g <glue path> <feature file path>`

Custom orders can be created by implementing the `PickleOrder` interface in the `cucumber.runtime.order` package. 

## Motivation and Context

Scenarios can be executed in random order. Aids in finding out if their is any dependency between scenarios.

## How Has This Been Tested?

Tests for `RuntimeOptions` are added in `RuntimeOptionsTest` for testing options in the command. `OrderTypeFactoryTest` tests the `OrderTypeFactory` class. `RandomOrderTypeTest` tests the `RandomOrderType` class. No test for `ReverseOrderType` as it will test the `reverse` method of `Collections` class.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue).
- [x ] New feature (non-breaking change which adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).

## Checklist:

- [x ] I've added tests for my code.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: core/src/main/java/cucumber/runtime/RuntimeOptions.java [hunk 1]
```diff
@@ -180,6 +184,10 @@ private void parse(List<String> args) {
                 parsedJunitOptions.addAll(asList(arg.substring("--junit,".length()).split(",")));
             } else if (arg.equals("--wip") || arg.equals("-w")) {
                 wip = true;
+            } else if (arg.equals("--order")) {
+            	orderTypeData.put(OrderType.TYPE_NAME, args.remove(0));
```
Comment (mpkorstanje) [inline core/src/main/java/cucumber/runtime/RuntimeOptions.java]: The names of the CLI commands are specific to the runtime options parser. It is not necessary for the `OrderType` to know what it is named on the CLI. It is okay to inline this string here.

File: core/src/main/java/cucumber/runtime/RuntimeOptions.java [hunk 2]
```diff
@@ -214,6 +222,10 @@ private void parse(List<String> args) {
             junitOptions.clear();
             junitOptions.addAll(parsedJunitOptions);
         }
+        if(orderTypeData.containsKey(OrderType.TYPE_NAME)) {
+        	orderType = OrderTypeFactory.getOrderType(orderTypeData);
+        	orderType.checkVariableValues(orderTypeData);
```
Comment (mpkorstanje) [inline core/src/main/java/cucumber/runtime/RuntimeOptions.java]: The exact syntax of CLI commands is also specific to the runtime options parser. Checking if the input is valid should be done here.

Additionally any time when `--count` is used it's argument should be a number and should non-negative. Tying the validation to the order type makes this more complex then needed. 

I think it might be better to make `--count` and `--order` independent variables and simply ignore `--count` when not relevant.

File: core/src/main/java/cucumber/runtime/order/OrderType.java [hunk 3]
```diff
@@ -0,0 +1,26 @@
+package cucumber.runtime.order;
+
+import java.util.List;
+import java.util.Map;
+
+import gherkin.events.PickleEvent;
+
+public abstract class OrderType {
+
+	public static final String NONE_ORDER_TYPE = "None";
```
Comment (mpkorstanje) [inline core/src/main/java/cucumber/runtime/order/OrderType.java]: Type names should be lower case or case insensitive.

File: core/src/main/java/cucumber/runtime/order/OrderType.java [hunk 4]
```diff
@@ -0,0 +1,26 @@
+package cucumber.runtime.order;
+
+import java.util.List;
+import java.util.Map;
+
+import gherkin.events.PickleEvent;
+
+public abstract class OrderType {
```
Comment (mpkorstanje) [inline core/src/main/java/cucumber/runtime/order/OrderType.java]: I think this class can be turned into an interface. The default implementation for `orderPickleEvents` is only used by `none` order.

File: core/src/main/java/cucumber/runtime/order/OrderTypeFactory.java [hunk 5]
```diff
@@ -0,0 +1,28 @@
+package cucumber.runtime.order;
+
+import java.util.Map;
+
+import cucumber.runtime.CucumberException;
+
+public class OrderTypeFactory {
```
Comment (mpkorstanje) [inline core/src/main/java/cucumber/runtime/order/OrderTypeFactory.java]: This mapping belongs in `RuntimeOptions`. Additionally I think the different orders, could be turned into an enum that implements/extends the `OrderType` interface/class.

File: core/src/main/java/cucumber/runtime/order/RandomOrderType.java [hunk 6]
```diff
@@ -0,0 +1,39 @@
+package cucumber.runtime.order;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+
+import cucumber.runtime.CucumberException;
+import gherkin.events.PickleEvent;
+
+public class RandomOrderType extends OrderType {
```
Comment (mpkorstanje) [inline core/src/main/java/cucumber/runtime/order/RandomOrderType.java]: This should be an Enum that implements/extends `OrderType`.

File: core/src/main/java/cucumber/runtime/order/RandomOrderType.java [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+	}
+
+	public RandomOrderType(int count) {
+		this.count = count;
+	}
+
+	@Override
+	public List<PickleEvent> orderPickleEvents(List<PickleEvent> pickleEvents) {	
+		Collections.shuffle(pickleEvents);
+		return pickleEvents.subList(0, ((count > pickleEvents.size() || count == 0) ?
```
Comment (mpkorstanje) [inline core/src/main/java/cucumber/runtime/order/RandomOrderType.java]: Nifty but way to hard to follow. 

I think this class is doing too much. It's both ordering and selecting pickles. It should only order pickles. By making `--order` and `--count` independent of each other you can remove this complexity all together.

File: core/src/main/java/cucumber/runtime/Runtime.java [hunk 8]
```diff
@@ -81,20 +81,25 @@ public void run() {
...
-                            runnerSupplier.get().runPickle(pickleEvent);
-                        }
-                    }));
+                	filteredEvents.add(pickleEvent);
                 }
             }
         }
+        
+        final List<Future<?>> executingPickles = new ArrayList<>();
+        for(final PickleEvent pickleEvent : runtimeOptions.getOrderType().orderPickleEvents(filteredEvents)) {
```
Comment (mpkorstanje) [inline core/src/main/java/cucumber/runtime/Runtime.java]: Ideally the order is injected into the `Runtime` via the builder. This allows us and other implementations to create `Runtime` instances without passing in CLI based `RuntimeOptions`.

This will in the long run let us separate the parsing of CLI options from executing pickles.

File: core/src/main/java/cucumber/runtime/order/RandomOrderType.java (hunk 6 above)
Comment (mpkorstanje) [reply core/src/main/java/cucumber/runtime/order/RandomOrderType.java]: https://stackoverflow.com/questions/2709593/why-would-an-enum-implement-an-interface

File: core/src/main/java/cucumber/runtime/Runtime.java (hunk 8 above)
Comment (mpkorstanje) [reply core/src/main/java/cucumber/runtime/Runtime.java]: For example:

```
        List<PickleEvent> allPickles = new ArrayList<>();
        for (CucumberFeature feature : features) {
            allPickles.addAll(feature.getPickles());
        }

        List<PickleEvent> orderedPickles = pickleOrder.order(allPickles);
        List<PickleEvent> filteredPickles = pickleFilter.filter(orderedPickles);

        final List<Future<?>> executingPickles = new ArrayList<>();
        for (PickleEvent filtredPickle : filteredPickles) {
            if (filters.matchesFilters(pickleEvent)) {
                executingPickles.add(executor.submit(new Runnable() {
                    @Override
                    public void run() {
                        runnerSupplier.get().runPickle(pickleEvent);
                    }
                }));
            }
        }
```

File: core/src/main/java/cucumber/runtime/RuntimeOptions.java (hunk 1 above)
Comment (grasshopper7) [reply core/src/main/java/cucumber/runtime/RuntimeOptions.java]: Will change it.

File: core/src/main/java/cucumber/runtime/RuntimeOptions.java (hunk 2 above)
Comment (grasshopper7) [reply core/src/main/java/cucumber/runtime/RuntimeOptions.java]: The `--count`  is used only for the `RandomOrderType` thus the check was specific to that type. That required the test for number etc to be done later.

Idea of making `--count` applicable for all types, to be used if required, does simplifies things. If so then the `orderTypeData` map is also redundant and can be removed. 

Later on if a future `OrderType` requires an additional variable from CLI it needs to be handled accordingly.

File: core/src/main/java/cucumber/runtime/order/OrderType.java (hunk 3 above)
Comment (grasshopper7) [reply core/src/main/java/cucumber/runtime/order/OrderType.java]: Will change it.

File: core/src/main/java/cucumber/runtime/order/RandomOrderType.java (hunk 7 above)
Comment (grasshopper7) [reply core/src/main/java/cucumber/runtime/order/RandomOrderType.java]: Yeah my bad. Though this logic needs to remain, as need to make sure that the `count` is less or equal to the size of the incoming PickleEvents to avoid exception.

File: core/src/main/java/cucumber/runtime/Runtime.java (hunk 8 above)
Comment (grasshopper7) [reply core/src/main/java/cucumber/runtime/Runtime.java]: To implement the above the existing iteration for the filters will need to be moved to the `Filters` class. Are you suggesting above that the pickles need to be ordered first and then filtered?

File: core/src/main/java/cucumber/runtime/order/RandomOrderType.java (hunk 6 above)
Comment (grasshopper7) [reply core/src/main/java/cucumber/runtime/order/RandomOrderType.java]: Will convert into enum.  No need of Factory.

File: core/src/main/java/cucumber/runtime/order/RandomOrderType.java (hunk 7 above)
Comment (mpkorstanje) [reply core/src/main/java/cucumber/runtime/order/RandomOrderType.java]: No worries!

File: core/src/main/java/cucumber/runtime/Runtime.java (hunk 8 above)
Comment (mpkorstanje) [reply core/src/main/java/cucumber/runtime/Runtime.java]: Yes. It'd be much easier to implement this with a Java8 `Stream` but we don't have those at the moment. So I don't mind too much where the iteration ends up.

**Comment by grasshopper7** [comment]
Implemented `OrderType` as an enumeration implementing `OrderPickleEvents` interface. 

Separated `--order` and `--count` CLI options to work independently. The `--order` logic is handled by `OrderType` enum. While the `--count` is handled by the `limitPickleEvents` method in `Filters` class. 

Updated the test classes accordingly.

**Comment by mpkorstanje** [comment]
Looks good! I'll need a bit more time to give this a proper once over.

File: core/src/main/java/cucumber/runtime/Runtime.java [hunk 9]
```diff
@@ -52,14 +53,16 @@
...
 
     public Runtime(final Plugins plugins,
                    final RuntimeOptions runtimeOptions,
                    final EventBus bus,
                    final Filters filters,
                    final RunnerSupplier runnerSupplier,
                    final FeatureSupplier featureSupplier,
-                   final ExecutorService executor) {
+                   final ExecutorService executor,
+                   final OrderType orderType) {
```
Comment (mpkorstanje) [inline core/src/main/java/cucumber/runtime/Runtime.java]: Did you mean to use `OrderPickleEvents` here?

File: core/src/main/java/cucumber/runtime/order/OrderType.java [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+		public List<PickleEvent> orderPickleEvents(List<PickleEvent> pickleEvents) {
+			Collections.shuffle(pickleEvents);
+			return pickleEvents;
+		}		
+	};
+
+
+     public static OrderType getOrderType(String name) {
+        for (OrderType orderType : OrderType.values()) {
+            if (name.equalsIgnoreCase(orderType.name())) {
```
Comment (mpkorstanje) [inline core/src/main/java/cucumber/runtime/order/OrderType.java]: Ideally we parse the different order names in runtime options too. This would avoid having to support both lower and uppercase options.
